From: Simon McVittie <smcv@collabora.com>
Date: Fri, 16 Apr 2021 10:46:01 +0100
Subject: transaction: Avoid a possible use-after-free

scan-build detected that mark_op_resolved() can be called with
op->resolved_commit == commit, in which case we incorrectly freed the
string before allocating the new copy.

Signed-off-by: Simon McVittie <smcv@collabora.com>
Forwarded: https://github.com/flatpak/flatpak/pull/4226
---
 common/flatpak-transaction.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/common/flatpak-transaction.c b/common/flatpak-transaction.c
index 1927498..3678fc1 100644
--- a/common/flatpak-transaction.c
+++ b/common/flatpak-transaction.c
@@ -2971,8 +2971,12 @@ mark_op_resolved (FlatpakTransactionOperation *op,
   g_assert (commit != NULL);
 
   op->resolved = TRUE;
-  g_free (op->resolved_commit); /* This is already set if we retry resolving to get a token, so free first */
-  op->resolved_commit = g_strdup (commit);
+
+  if (op->resolved_commit != commit)
+    {
+      g_free (op->resolved_commit); /* This is already set if we retry resolving to get a token, so free first */
+      op->resolved_commit = g_strdup (commit);
+    }
 
   if (sideload_path)
     op->resolved_sideload_path = g_object_ref (sideload_path);
